src/cpu/core_dyn_x86.cpp

```diff
@@ -263,25 +275,36 @@ static void copy_normal_fpu_to_dh()
 	dyn_dh_fpu.state.sw  = FPU_GetSW();
 	dyn_dh_fpu.state.cw  = FPU_GetCW();
 	dyn_dh_fpu.state.tag = FPU_GetTag();
-	FPU_SavePRegs(&dyn_dh_fpu.state.st_reg[0][0]);
+	FPU_GetPRegsTo(dyn_dh_fpu.state.st_reg);
 }
 
+static Bits sync_fpu_and_run_normal_core() noexcept
+{
+	if (dyn_dh_fpu.state_used) {
+		// If the host's FPU was used by a recompiled code block, we
+		// copy its state into dyn_dh_fpu.state and from there into the
+		// normal core's emulated FPU
+		copy_dh_fpu_to_normal();
+	}
+
+	auto ret = CPU_Core_Normal_Run();
+
+	// Always copy the normal core's emulated FPU state back to the dynamic
+	// core's state on leaving the normal core, as the normal core's FPU
+	// usage cannot be determined
+	copy_normal_fpu_to_dh();
+	return ret;
+}
+#	else
+static Bits sync_fpu_and_run_normal_core() noexcept
+{
+	return CPU_Core_Normal_Run();
+}
+#	endif
+
 Bits CPU_Core_Dyn_X86_Run() noexcept
 {
 	ZoneScoped;
-	// helper class to auto-save DH_FPU state on function exit
-	class auto_dh_fpu {
```

Comment:
No you can't delete this either.  This makes sure whenever we leave CPU_Core_Dyn_X86_Run(),  we keep a snapshot of the dynrec's FPU state so later we can resume.   I would also keep the destructor pattern because this function has many exits, remember to manually saving states at each place is going to be a maintenance burden.